### ISSUE TITLE
GRPC Document API benchmark - 3.0.0 Experimental feature release

### ISSUE THREAD

**Issue by amberzsy** [opensearch-project/OpenSearch#16711]
### Please describe the end goal of this project

based on [GRPC document API](https://github.com/opensearch-project/OpenSearch/issues/16784) implementation, perform benchmark on varies document apis with different scenarios.

### Supporting References

https://github.com/opensearch-project/OpenSearch/issues/15190

### Issues

- [x] Bulk API benchmark based on https://github.com/opensearch-project/OpenSearch/issues/16784

### Related component

Indexing

**Comment by amberzsy** [comment]
assigned to @karenyrx. WIP - expect result by this Friday.

**Comment by karenyrx** [comment]
## Results
[Latency benchmark results](https://docs.google.com/document/d/e/2PACX-1vQnKPkS8f9HPc4I1ioMVXCDUEXavi5dTwpp0V8J8hc2OjKh1Q7w3EiT_z7UD0G0-hbxLPRbhDKV3uND/pub) from testing with a single data-node cluster and small GRPC payload: 

<img width="511" alt="Screenshot 2025-01-02 at 4 20 12 PM" src="https://github.com/user-attachments/assets/b0db236c-e6c4-4f11-a9f6-24f64bbfb9fd" />

## Summary
1. GRPC latency of the first request is 1.2 - 1.25x  higher than HTTP.
2. GRPC latency of the subsequent requests is 1.5 - 5x lower than HTTP. 
3. For both GRPC and HTTP, CPU, memory, IO usage were similar, based on internal dashboard observations. 

## Next steps: 
1. Confirm where the improvements are arising from: HTTP2, or Protobuf, or something else. To do so:
    i)  Instrument codebase with detailed metrics to breakdown granular latency of GRPC vs. HTTP internal ops.
    ii) Confirm if results align with HTTP2 + JSON benchmarking results  (to be performed by @reta )
2. Test with a larger payload (e.g. 5MB request) to gain more confidence 
3. Submit PR for Protobuf and GRPC support for Bulk endpoint in Opensearch  (in parallel with step 1 and 2)

**Comment by mgodwan** [comment]
@karenyrx Thanks for the benchmarks. The idea is indeed promising.
I would like to understand and discuss the proposal around how you're thinking to map document schema using protobuf, and how dynamic mappings would work for such scenarios.

**Comment by karenyrx** [comment]
> I would like to understand and discuss the proposal around how you're thinking to map document schema using protobuf, and how dynamic mappings would work for such scenarios.

@mgodwan This is out of scope for now.  For now we plan to translate APIs defined by the API specification. In the future, we do plan to explore having well-defined document schemas to take full advantage of the binary protocol and structured schema.

**Comment by karenyrx** [comment]
## Bulk API: Large Payload Benchmarking with gRPC vs HTTP 

### Key findings:

- gRPC consistently provides lower latency compared to HTTP at both p50 and p99.9 levels.
- HTTP shows slightly better performance at p99.9 latency for the first request only.
- The latency improvements arise from the request-side, while the response side has slightly longer latency compared to HTTP, but the net latency gains is still favorable for GRPC. 


### Test Setup

- **100 index + 100 update + 100 delete** operations per bulk request.
- Payload sizes tested: **0.5MB – 4.05MB**
- RPS tested: **1 – 10**
- Nodes: **1 data node (24 cores, 64GB mem), 3 master nodes**
- gRPC message size limit: **~4MB**
- Ballast & Muttley used for load generation and metrics.
- Opensearch version: 3.0

### Latency Results
#### Trial 1: 

##### P99.9 Latency vs Payload Size

![Image](https://github.com/user-attachments/assets/bb7365a8-2e74-4f6f-860a-a15c79ad06aa)

##### P50 Latency vs Payload Size

![Image](https://github.com/user-attachments/assets/a13a64f3-7927-458f-b50a-c1d14206d56c)

##### Raw data

| Payload Size (MB) | HTTP P99.9 (ms) | gRPC P99.9 (ms) | HTTP P50 (ms) | gRPC P50 (ms) |
|-------------------|------------------|------------------|----------------|----------------|
| 0.5               | 123              | 80.8             | 90.8           | 55             |
| 1.0               | 169              | 128              | 147            | 103            |
| 2.0               | 213              | 156              | 175            | 108            |
| 3.0               | 217              | 196              | 198            | 157            |
| 4.0               | 352              | 324              | 311            | 261            |

#### Trial 2: 

##### P99.9 Latency vs Payload Size

![Image](https://github.com/user-attachments/assets/8de7a4ef-ca3a-4e1f-91e4-b0b61f0b0726)

##### P50 Latency vs Payload Size

![Image](https://github.com/user-attachments/assets/b2c2ca0e-c93c-401e-b33a-b086d61926d3)

##### Raw data
|   Payload Size (MB) |   HTTP P99.9 (ms) |   gRPC P99.9 (ms) |   HTTP P50 (ms) |   gRPC P50 (ms) |
|--------------------:|------------------:|------------------:|----------------:|----------------:|
|                0.5  |               108 |              65.8 |            81.7 |            51.7 |
|                1    |               139 |             110   |           101   |            92.6 |
|                2    |               186 |             151   |           153   |           104   |
|                3    |               223 |             189   |           201   |           156   |
|                4    |               262 |             240   |           248   |           204   |



### Latency Breakdown 

### First Request (P99)

| Phase               | HTTP Latency (ms) | gRPC Latency (ms) |
|---------------------|-------------------|--------------------|
| Request Conversion  | 64                | 10                 |
| Response Conversion | 4                 | 35                 |
| **Total**           | **68**            | **45**             |

### Second Request (P99)
| Phase               | HTTP Latency (ms) | gRPC Latency (ms) |
|---------------------|-------------------|--------------------|
| Request Conversion  | 42                | 3                  |
| Response Conversion | 2                 | 13                 |
| **Total**           | **44**            | **16**             |


### Conclusion
#### GRPC vs. HTTP Latency
- GRPC overall has lower or comparable latency as HTTP

#### First vs. Subsequent Requests
- gRPC has a warm-up cost, making its first request slower at p99.9 than HTTP.
- Subsequent requests are consistently faster with gRPC.

#### Request-side vs. response-side latency
- Request-side latency improves, but response-side slightly worsens. The net gain is still GRPC having lower latency.
